relay-sampling/src/evaluation.rs

```diff
@@ -24,13 +28,116 @@ fn pseudo_random_from_uuid(id: Uuid) -> f64 {
     generator.sample(dist)
 }
 
+/// Increments the reservoir count for a given rule in redis.
+///
+/// - INCR docs: [`https://redis.io/commands/incr/`]
+/// - If the counter doesn't exist in redis, a new one will be inserted.
+#[cfg(feature = "redis")]
+fn increment_redis_reservoir_count(
+    redis: &Arc<RedisPool>,
+    org_id: u64,
+    rule_id: RuleId,
+) -> anyhow::Result<i64> {
+    let mut command = relay_redis::redis::cmd("INCR");
+
+    let key = format!("bias:{}:{}", org_id, rule_id);
+    command.arg(key);
+
+    Ok(command.query(&mut redis.client()?.connection()?)?)
+}
+
+/// The amount of matches for each reservoir rule in a given project.
+pub type ReservoirCounters = Arc<Mutex<BTreeMap<RuleId, i64>>>;
+
+/// Utility for evaluating reservoir-based sampling rules.
+///
+/// A "reservoir limit" rule samples every match until its limit is reached, after which
+/// the rule is disabled.
+///
+/// This utility uses a dual-counter system for enforcing this limit:
+///
+/// - Local Counter: Each relay instance maintains a local counter to track sampled events.
+///
+/// - Redis Counter: For processing relays, a Redis-based counter provides synchronization
+///   across multiple relay-instances. When incremented, the Redis counter returns the current global
+///   count for the given rule, which is then used to update the local counter.
+#[derive(Debug)]
+pub struct ReservoirEvaluator {
+    counters: ReservoirCounters,
+    #[cfg(feature = "redis")]
+    redis_pool: Option<Arc<RedisPool>>,
+    #[cfg(feature = "redis")]
+    org_id: Option<u64>,
+}
+
+impl ReservoirEvaluator {
+    /// Constructor for [`ReservoirEvaluator`].
+    pub fn new(counters: ReservoirCounters) -> Self {
+        Self {
+            counters,
+            #[cfg(feature = "redis")]
+            org_id: None,
+            #[cfg(feature = "redis")]
+            redis_pool: None,
+        }
+    }
+
+    /// Sets the Redis pool and organiation ID for the [`ReservoirEvaluator`].
+    ///
+    /// These values are needed to synchronize with Redis.
+    #[cfg(feature = "redis")]
+    pub fn set_redis(mut self, org_id: u64, redis_pool: Arc<RedisPool>) -> Self {
+        self.org_id = Some(org_id);
+        self.redis_pool = Some(redis_pool);
+        self
+    }
+
+    /// Evaluates a reservoir rule, returning `true` if it should be sampled.
+    pub fn evaluate(&self, rule: RuleId, limit: i64) -> bool {
+        // If the mutex is already locked, we abort the match, for performance reasons.
+
+        let Ok(mut map_guard) = self.counters.try_lock() else {
```

Comment:
thoughts on creating a metric here? We assume that we will almost always get the match, but we can't be sure.

perhaps it could crash with `fn remove_expired_reservoir_rules` when a new project config is received.

Reply:
You can run a proper `lock` here. Since we're in sync context, we're allowed to block now. However, I'd split this method into three logical sections: 

 1. (locked) check the cached entry and return early if we've exceeded our budget
 2. (unlocked) run the redis increment and obtain a new count.
 3. (locked) write the value returned from redis back into the cache
 
 there can be a fast-path if redis is not configured, in which case you just lock once, check, and increment.